keps/sig-architecture/5241-beta-featuregate-promotion-requirements/README.md

```diff
@@ -0,0 +1,246 @@
+We must avoid making every production cluster into unstable or incomplete feature testing clusters.
+Even feature gates that make flags accessible, but require a secondary configuration to use must be
+stable, because it is unrealistic to expect everyone to understand the graduation stages of various flags
+for each release: the only stages that really matter are "takes enabling an explicit alpha feature gate"
+and "my production cluster accepts this as valid by default".
+
+### Goals
+
+* Features gates must include all functional, security, and testing requirements along with
+  resolving all issues and gaps identified prior to being enabled by default.
+* The only valid GA criteria are “all issues and gaps identified as feedback during beta are resolved”.
+
+### Non-Goals
+
+* Changing beta APIs off by default rules.
+* Change the imperfect mechanisms we have for API evolution.
+
+## Proposal
+
+Kubernetes feature gates have three levels: GA, Beta, and Alpha.
+1. GA means that a feature gate is unconditionally enabled in all production kubernetes clusters and
+   that feature cannot be disabled.
+2. Beta means that a feature gate is usually enabled in all production Kubernetes clusters by default
+   and that feature can be disabled.
+   Exceptions exist for entirely new APIs and some node features, but this broadly the case.
+3. Alpha means that a feature gate is disabled in all production Kubernetes clusters by default and
+   can be optionally enabled by setting a `--feature-gate` command line argument.
+
+Making the jump to GA (cannot be disabled), without actual field experience is irresponsible.
+The first time we take a feature gate enabled by default in production Kubernetes clusters, we must
+have a way to disable the feature in case of unexpected stability, performance, or security issues.
+
+Enabling incomplete features in production Kubernetes clusters by default is irresponsible.
+Features that are known to be incomplete naturally bring with them additional stability, performance, and security issues.
+Once a feature has been enabled in a production Kubernetes cluster by default, adding to it carries
+greater risk to upgrading clusters and the ecosystem.
+The feature can easily have become relied upon by workloads and other platform extensions.
+If an accident happens in adding those capabilities with stability, performance, and security the
+cost to disable those features in a cluster becomes significantly greater and breaks existing
+clusters, workloads and use-cases.
+This posture makes upgrades higher risk than necessary.
+
+To balance these concerns, we are changing how we evaluate Beta and GA stability criteria.
+The only valid GA criteria are “all issues and gaps identified as feedback during beta are resolved”.
+Promotion from Beta to GA must be zero-diff for the release.
+This means that Beta criteria must include all functional, security, and testing requirements along
```

Comment:
I think we should also add monitoring and I know that @lmktfy has mentioned satisfactory documentation in beta before we promote to GA.

Reply:
> I think we should also add monitoring and I know that @lmktfy has mentioned satisfactory documentation in beta before we promote to GA.

updating